server/entity/firework.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+	"math/rand"
+)
+
+// Firework is an item (and entity) used for creating decorative explosions, boosting when flying with elytra, and
+// loading into a crossbow as ammunition.
+type Firework struct {
+	transform
+
+	yaw, pitch float64
+	firework   item.Firework
+
+	owner world.Entity
+
+	c *MovementComputer
+
+	ticks int
+	close bool
+}
+
+// NewFirework ...
+func NewFirework(pos mgl64.Vec3, yaw, pitch float64, firework item.Firework) *Firework {
+	f := &Firework{
+		yaw:      yaw,
+		pitch:    pitch,
+		firework: firework,
+		c:        &MovementComputer{},
+		ticks:    int(firework.RandomizedDuration().Milliseconds() / 50),
+	}
+	f.transform = newTransform(f, pos)
+	f.vel = mgl64.Vec3{rand.Float64() * 0.001, 0.05, rand.Float64() * 0.001}
+	return f
+}
+
+// Name ...
+func (f *Firework) Name() string {
+	return "Firework Rocket"
+}
+
+// EncodeEntity ...
+func (f *Firework) EncodeEntity() string {
+	return "minecraft:fireworks_rocket"
+}
+
+// BBox ...
+func (f *Firework) BBox() cube.BBox {
+	return cube.BBox{}
+}
+
+// Firework returns the underlying item.Firework of the Firework.
+func (f *Firework) Firework() item.Firework {
+	return f.firework
+}
+
+// Rotation ...
+func (f *Firework) Rotation() (float64, float64) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return f.yaw, f.pitch
+}
+
+// Tick ...
+func (f *Firework) Tick(w *world.World, current int64) {
+	if f.close {
+		_ = f.Close()
+		return
+	}
+
+	f.mu.Lock()
+	f.vel[0] *= 1.15
+	f.vel[1] += 0.04
+	f.vel[2] *= 1.15
+	m := f.c.TickMovement(f, f.pos, f.vel, f.yaw, f.pitch)
+	f.pos, f.vel, f.yaw, f.pitch = m.pos, m.vel, m.yaw, m.pitch
+	f.mu.Unlock()
+
+	m.Send()
+
+	if m.pos[1] < float64(w.Range()[0]) && current%10 == 0 {
+		f.close = true
+		return
+	}
+
+	f.ticks--
+	if f.ticks < 0 {
+		explosions := f.Firework().Explosions
+		for _, v := range w.Viewers(m.pos) {
+			v.ViewEntityAction(f, FireworkParticleAction{})
+		}
+		for _, explosion := range explosions {
+			if explosion.Shape == item.FireworkShapeHugeSphere() {
+				w.PlaySound(m.pos, sound.FireworkHugeBlast{})
+			} else {
+				w.PlaySound(m.pos, sound.FireworkBlast{})
+			}
+			if explosion.Twinkle {
+				w.PlaySound(m.pos, sound.FireworkTwinkle{})
+			}
+		}
+
+		if len(explosions) > 0 {
+			force := float64(len(explosions)*2) + 5.0
+			for _, e := range w.EntitiesWithin(f.BBox().Translate(m.pos).Grow(5.25), func(e world.Entity) bool {
+				l, living := e.(Living)
+				return !living || l.AttackImmune()
+			}) {
+				pos := e.Position()
+				dist := m.pos.Sub(pos).Len()
+				if dist > 5.0 {
+					// The maximum distance allowed is 5.0 blocks.
+					continue
+				}
+				if _, ok := trace.Perform(m.pos, pos, w, e.BBox().Grow(0.3), func(world.Entity) bool {
+					return true
+				}); ok {
```

Comment:
This `func(world.Entity) bool` can simply be `nil`.

Reply:
Don't think so? We need to ignore all entities, not allow all entities.

Reply:
Ah I see, my bad!